OpenMP inscan exclusive fails to get parsed correctly with clang-22 if it is used for a small prefix loop contained in a longer function

Hi, i am currently writing a library that provides tensor and linear algebra operations on gpu and cpu, as well as distributed with the message passing interface. The code can be found here:

https://github.com/bschulz81/AcceleratedLinearAlgebra/

It is a bit more complex that a simple test case.

Unfortunately, I came across a compiler bug  in clang-22 that I don't know how to reduce,

It occurs only if the inscan exclusive directive of OpenMp is used in a longer function.

So at first I give an intro how to compile the test case.

In 

https://github.com/bschulz81/AcceleratedLinearAlgebra/ 

there is a Cmakelists.txt.

If one comments out in this Cmakelists.txt the line:

```cmake
SET (CMAKE_CXX_COMPILER "/usr/bin/g++-16" CACHE STRING "C++ compiler" FORCE)
SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fopenmp -foffload=nvptx-none -fno-stack-protector -O1 -Wall")
```

and removes the comments from

```cmake
#SET (ENV{LIBRARY_PATH} "/usr/lib64/nvptx64-nvidia-cuda/:$ENV{LIBRARY_PATH}")
#SET (CMAKE_CXX_COMPILER /usr/lib/llvm/22/bin/clang++ CACHE STRING "C++ compiler" FORCE)
#SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++23 -fopenmp -fopenmp-targets=nvptx64-nvidia-cuda -Wall")
```

then 

cmake .

and 

make 

will compile several test applications with clang-22.

Now In the file

https://github.com/bschulz81/AcceleratedLinearAlgebra/blob/main/datablock_mpifunctions.h

there is a function that scatters a tensor on several OpenMP nodes. 

In this function, there is line 1818, which executes the following loop on host (i.e. not on any target device):

```c
for(size_t i = 0; i < local_blocks; i++)
{
    recv_db.pblock_offsets[i] = total_recv_elems;
    total_recv_elems += blocks[i].blocksize;
}
```

Unchanged, this will compile with clang-22

But this is of course ideal for a prefix sum with openmp inscan in openmp. 

So one may replace this loop with:

```c
size_t total_recv_elems = 0;
#pragma omp parallel for simd reduction(inscan, +: total_recv_elems)
for(size_t i = 0; i < local_blocks; i++)
{
recv_db.pblock_offsets[i] = total_recv_elems;

#pragma omp scan exclusive(total_recv_elems)
total_recv_elems += blocks[i].blocksize;
}
```

I then recieve the following error from clang:

/home/benni/projects/arraylibrary/openmp/datablock_mpifunctions.h:1792:51: error: the inscan reduction list item must appear as a list item in an 'inclusive' or 'exclusive' clause on an inner 'omp scan' directive 1792 | #pragma omp parallel for reduction(inscan, +: total_recv_elems)

The message makes no sense as the loop is basically textbook OpenMp.

Furthermore, gcc-16 will compile the changed file without any problem.

I then looked for an example with inscan exclusive from the OpenMP site:

https://passlab.github.io/Examples/contents/Chap_data_environment/10_scan_Directive.html

I changed the variable names to avoid any interference with my code, and inserted this  self contained code block from the OpenMP site at line 1817 into the file:

https://github.com/bschulz81/AcceleratedLinearAlgebra/blob/main/datablock_mpifunctions.h

```c
{
const int F=100;

int u[F], v[F];
int w= 0;

// initialization
for (int k = 0; k < F; k++)
u[k] = k + 1;

// u[k] is not included in the computation of producing
// results in v[k]
#pragma omp parallel for simd reduction(inscan,+: w)
for (int k = 0; k < F; k++) {
v[k] = w;
#pragma omp scan exclusive(w)
w += u[k];
}
}
```

I again recieved this error from clang, but now for the example from OpenMP which is proven to be correct.

/home/benni/projects/arraylibrary/openmp/datablock_mpifunctions.h:1833:40: error: the list item must appear in 'reduction' clause with the 'inscan' modifier of the parent directive
 1833 |             #pragma omp scan exclusive(w)
      |                                        ^
/home/benni/projects/arraylibrary/openmp/datablock_mpifunctions.h:1829:59: error: the inscan reduction list item must appear as a list item in an 'inclusive' or 'exclusive' clause on an inner 'omp scan' directive
 1829 |         #pragma omp parallel for simd reduction(inscan,+: w)
      |                      

The point is that the example compiles with clang when it is used in a trivial main function and in a program with a single cpp file. 

But when I insert that block in my more involved function for tensor scattering, clang refuses to compile and claims I use openmp wrongly.

It should not happen, that a self contained code block, which is verified to be correct and which does not interact with the surrounding code in any way, does does not get parsed correctly anymore when it is used in a larger application.

So it seems there are problems with the inscan directive and clang when it is used within larger functions.

Additionally, gcc would compile this without warnings, and use the inscan directive correctly.

Since the problem only appears if the inscan loop is inserted in a larger function, I don't know how to reduce the test case, sorry..

--- a/clang/lib/Sema/SemaOpenMP.cpp
+++ b/clang/lib/Sema/SemaOpenMP.cpp
@@ -20498,6 +20498,10 @@ static bool actOnOMPReductionKindClause(
         (DeclareReductionRef.isUnset() ||
          isa<UnresolvedLookupExpr>(DeclareReductionRef.get()))) {
       RD.push(RefExpr, DeclareReductionRef.get());
+      // Handle non-dependent inscan reduction variables in dependent contexts.
+      if (RD.RedModifier == OMPC_REDUCTION_inscan)
+        Stack->addDSA(D, RefExpr->IgnoreParens(), OMPC_reduction, nullptr,
+                      RD.RedModifier, ASE || OASE);
       continue;
     }
     if (BOK == BO_Comma && DeclareReductionRef.isUnset()) {

--- a/clang/test/OpenMP/scan_inscan_template_nondependent.cpp
+++ b/clang/test/OpenMP/scan_inscan_template_nondependent.cpp
@@ -1,0 +1,63 @@
+// RUN: %clang_cc1 -verify -fopenmp -std=c++17 %s
+// RUN: %clang_cc1 -verify -fopenmp-simd -std=c++17 %s
+// expected-no-diagnostics
+
+// Regression test for https://github.com/llvm/llvm-project/issues/191549
+// Inscan reductions with non-dependent types in function templates were
+// incorrectly rejected because the DSA stack was not populated in
+// dependent contexts.
+
+template <typename T>
+void exclusive_in_template() {
+    int a[100], b[100];
+    int sum = 0;
+    #pragma omp parallel for reduction(inscan, +: sum)
+    for (int i = 0; i < 100; i++) {
+        a[i] = sum;
+        #pragma omp scan exclusive(sum)
+        sum += b[i];
+    }
+}
+
+template <typename T>
+void inclusive_in_template() {
+    int a[100], b[100];
+    int sum = 0;
+    #pragma omp parallel for reduction(inscan, +: sum)
+    for (int i = 0; i < 100; i++) {
+        sum += b[i];
+        #pragma omp scan inclusive(sum)
+        a[i] = sum;
+    }
+}
+
+template <typename T>
+void for_simd_inscan_in_template() {
+    int a[100], b[100];
+    int sum = 0;
+    #pragma omp parallel for simd reduction(inscan, +: sum)
+    for (int i = 0; i < 100; i++) {
+        a[i] = sum;
+        #pragma omp scan exclusive(sum)
+        sum += b[i];
+    }
+}
+
+template <typename T>
+void simd_inscan_in_template() {
+    int a[100], b[100];
+    int sum = 0;
+    #pragma omp simd reduction(inscan, +: sum)
+    for (int i = 0; i < 100; i++) {
+        a[i] = sum;
+        #pragma omp scan exclusive(sum)
+        sum += b[i];
+    }
+}
+
+void instantiate() {
+    exclusive_in_template<int>();
+    inclusive_in_template<int>();
+    for_simd_inscan_in_template<int>();
+    simd_inscan_in_template<int>();
+}

--- a/openmp/runtime/test/worksharing/for/omp_scan_inscan_template.cpp
+++ b/openmp/runtime/test/worksharing/for/omp_scan_inscan_template.cpp
@@ -1,0 +1,73 @@
+// RUN: %libomp-cxx-compile-and-run | FileCheck %s
+
+// Regression test for https://github.com/llvm/llvm-project/issues/191549
+// Inscan reductions with non-dependent types inside function templates were
+// rejected by Clang because the DSA stack was not populated in dependent
+// contexts.  This test verifies runtime correctness of both exclusive and
+// inclusive scans in template functions.
+
+#include <cstdlib>
+#include <numeric>
+#include <stdio.h>
+
+#define N 100
+
+template <typename T> bool test_exclusive(T *a, const T *b) {
+  int sum = 0;
+#pragma omp parallel for reduction(inscan, + : sum)
+  for (int i = 0; i < N; i++) {
+    a[i] = sum;
+#pragma omp scan exclusive(sum)
+    sum += b[i];
+  }
+
+  for (int i = 0, prefix = 0; i < N; i++) {
+    if (a[i] != prefix)
+      return false;
+    prefix += b[i];
+  }
+  return true;
+}
+
+template <typename T> bool test_inclusive(T *a, const T *b) {
+  int sum = 0;
+#pragma omp parallel for reduction(inscan, + : sum)
+  for (int i = 0; i < N; i++) {
+    sum += b[i];
+#pragma omp scan inclusive(sum)
+    a[i] = sum;
+  }
+
+  for (int i = 0, prefix = 0; i < N; i++) {
+    prefix += b[i];
+    if (a[i] != prefix)
+      return false;
+  }
+  return true;
+}
+
+int main() {
+  bool success = true;
+
+  int a[N], b[N];
+  std::iota(b, b + N, 1);
+
+  if (test_exclusive<int>(a, b)) {
+    printf("exclusive: PASS\n");
+  } else {
+    printf("exclusive: FAIL\n");
+    success = false;
+  }
+
+  if (test_inclusive<int>(a, b)) {
+    printf("inclusive: PASS\n");
+  } else {
+    printf("inclusive: FAIL\n");
+    success = false;
+  }
+
+  return success ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+
+// CHECK: exclusive: PASS
+// CHECK: inclusive: PASS

